modules/login/php/ajaxlogin.class.inc

```diff
@@ -0,0 +1,405 @@
+        } else if (isset($values['command']) && $values['command'] == 'login') {
+            // login - user attempting to login.
+            $response = $this->login($values);
+        } else if (isset($values['command']) && $values['command'] == 'expired') {
+            // login - expired password update
+            $response = $this->expiredPassword($values);
+        } else {
+            $response['status'] = 'error';
+        }
+        return $response;
+    }
+    /**
+     * Initialize setup.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function initializeSetup(array $values) : array
+    {
+        // Login Setup.
+        $loginData = array();
+        $login     = $_SESSION['State']->getProperty('login');
+        // Previous login failed.
+        if (!empty($login->_lastError)) {
+            $loginData['error'] = $login->_lastError;
+        }
+        $config      =& \NDB_Config::singleton();
+        $study_links = $config->getExternalLinks('Studylinks');
+        foreach ($study_links as $label => $url) {
+            $windowName           = md5($url);
+            $loginData['links'][] = array(
+                                     'url'        => $url,
+                                     'label'      => $label,
+                                     'windowName' => $windowName,
+                                    );
+        }
+        $loginData['description'] = html_entity_decode(
+            $config->getSetting('StudyDescription')
+        );
+        $loginData['title']       = $config->getSetting('title');
+        $loginData['logo']        = $config->getSetting('studylogo');
+        $values['login']          = $loginData;
+
+        // Request Account Setup.
+        $requestAccountData = array();
+        $reCAPTCHAPrivate   = $config->getSetting('reCAPTCHAPrivate');
+        $reCAPTCHAPublic    = $config->getSetting('reCAPTCHAPublic');
+        // reCAPTCHA if both private and public keys are set
+        if ($reCAPTCHAPrivate && $reCAPTCHAPublic) {
+            $requestAccountData['captcha'] = $reCAPTCHAPublic; // captcha key.
+        }
+        $requestAccountData['site'] = \Utility::getSiteList();
+        $values['requestAccount']   = $requestAccountData;
+
+        return $values;
+    }
+
+    /**
+     * Reset password.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to phishers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
+
+        // Check if valid user.
+        $email = '';
+        try {
+            $email = $user->getData('Email');
+            if (empty($email) || empty($user)) {
```

Comment:
The \User::singleton() will never return a value that equates to `empty` so this should be changed.

Reply:
@johnsaigle I guess it should be a is_null check for $user.

Reply:
It also can't return null if you check the signature in `php/libraries/User.class.inc`. It either returns a User object (singleton) with all the correct data or an AnonymousUser object.

Reply:
@johnsaigle okay thanks i'll remove it.